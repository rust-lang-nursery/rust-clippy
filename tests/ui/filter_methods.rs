#![warn(clippy::all, clippy::pedantic)]
#![allow(clippy::missing_docs_in_private_items)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::unnecessary_filter_map)]

fn main() {
    let options: Vec<Option<i8>> = vec![Some(5), None, Some(6)];
    let results = ["1", "lol", "3", "NaN", "5"];

    // validate iterator of values triggers filter_map + map lint
    let _: Vec<_> = vec![5_i8; 6].into_iter()
        .filter_map(|x| x.checked_mul(2))
        .map(|x| x.checked_mul(2))
        .collect();

    // validate iterator of options triggers filter + map lint
    let _: Vec<i8> = options.clone().into_iter()
        .filter(|x| x.is_some())
        .map(|x| x.unwrap())
        .collect();

    // validate iterator of options triggers filter + flat_map lint
    let _: Vec<Option<i8>> = std::iter::repeat(options.clone())
        .take(5)
        .map(|v| Some(v.into_iter()))
        .filter(|x| x.is_some())
        .flat_map(|x| x.unwrap())
        .collect();

    // validate iterator of results triggers filter + map lint
    let _: Vec<i8> =  results.iter()
        .map(|s| s.parse())
        .filter(|s| s.is_ok())
        .map(|s| s.unwrap())
        .collect();

    // validate iterator of values **does not** trigger filter + map lint
    let _: Vec<String> =  results.iter()
        .filter(|s| s.len() > 3)
        .map(|s| format!("{}{}", s, s))
        .collect();

    // validate iterator of values **does not** trigger filter + flat_map lint
    let _: String = results.iter()
        .filter(|i| i.len() > 1)
        .flat_map(|s| s.chars())
        .collect();

    // validate filter_map + flat_map **does not** trigger linter
    let _: Vec<_> = vec![5_i8; 6]
        .into_iter()
        .filter_map(|x| x.checked_mul(2))
        .flat_map(|x| x.checked_mul(2))
        .collect();
}
