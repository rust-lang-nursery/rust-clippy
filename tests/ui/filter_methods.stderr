error: redundant closure found
  --> $DIR/filter_methods.rs:19:14
   |
LL |         .map(|x| x.unwrap())
   |              ^^^^^^^^^^^^^^ help: remove closure as shown: `std::option::Option::unwrap`
   |
   = note: `-D clippy::redundant-closure-for-method-calls` implied by `-D warnings`

error: called `filter(p).map(q)` on an `Iterator<Option<_>>`. Consider calling `.filter_map(..)` instead.
  --> $DIR/filter_methods.rs:18:10
   |
LL |           .filter(|x| x.is_some())
   |  __________^
LL | |         .map(|x| x.unwrap())
   | |____________________________^
   |
   = note: `-D clippy::filter-map` implied by `-D warnings`

error: redundant closure found
  --> $DIR/filter_methods.rs:18:17
   |
LL |         .filter(|x| x.is_some())
   |                 ^^^^^^^^^^^^^^^ help: remove closure as shown: `std::option::Option::is_some`

error: redundant closure found
  --> $DIR/filter_methods.rs:27:19
   |
LL |         .flat_map(|x| x.unwrap())
   |                   ^^^^^^^^^^^^^^ help: remove closure as shown: `std::option::Option::unwrap`

error: redundant closure found
  --> $DIR/filter_methods.rs:26:17
   |
LL |         .filter(|x| x.is_some())
   |                 ^^^^^^^^^^^^^^^ help: remove closure as shown: `std::option::Option::is_some`

error: redundant closure found
  --> $DIR/filter_methods.rs:34:14
   |
LL |         .map(|s| s.unwrap())
   |              ^^^^^^^^^^^^^^ help: remove closure as shown: `std::result::Result::unwrap`

error: called `filter(p).map(q)` on an `Iterator<Result<_>>`. Consider calling `.filter_map(..)` instead.
  --> $DIR/filter_methods.rs:33:10
   |
LL |           .filter(|s| s.is_ok())
   |  __________^
LL | |         .map(|s| s.unwrap())
   | |____________________________^

error: redundant closure found
  --> $DIR/filter_methods.rs:33:17
   |
LL |         .filter(|s| s.is_ok())
   |                 ^^^^^^^^^^^^^ help: remove closure as shown: `std::result::Result::is_ok`

error: aborting due to 8 previous errors

